pkg/ruler/ruler.go

```diff
@@ -1320,3 +1333,49 @@ func returnByState(requestState string, alertState string) bool {
 func returnByHealth(requestHealth string, ruleHealth string) bool {
 	return requestHealth == "" || requestHealth == ruleHealth
 }
+
+func parseMatchersParam(matchers []string) ([][]*labels.Matcher, error) {
+	var matcherSets [][]*labels.Matcher
+	for _, s := range matchers {
+		matchers, err := parser.ParseMetricSelector(s)
+		if err != nil {
+			return nil, err
+		}
+		matcherSets = append(matcherSets, matchers)
+	}
+
+OUTER:
+	for _, ms := range matcherSets {
+		for _, lm := range ms {
+			if lm != nil && !lm.Matches("") {
+				continue OUTER
+			}
+		}
+		return nil, errors.New("match[] must contain at least one non-empty matcher")
+	}
+	return matcherSets, nil
+}
+
+func matches(l labels.Labels, matchers ...*labels.Matcher) bool {
+	for _, m := range matchers {
+		if v := l.Get(m.Name); !m.Matches(v) {
+			return false
+		}
+	}
+	return true
+}
+
+// matchesMatcherSets ensures all matches in each matcher set are ANDed and the set of those is ORed.
+func matchesMatcherSets(matcherSets [][]*labels.Matcher, l labels.Labels) bool {
+	if len(matcherSets) == 0 {
+		return true
+	}
+
+	var ok bool
+	for _, matchers := range matcherSets {
+		if matches(l, matchers...) {
+			ok = true
```

Comment:
It is unnecessary to introduce `ok` here as it keeps checking `matchers` for all `matcherSets`.
We can return early if at least one matches.